core/unit/name.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package unit
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/juju/juju/internal/errors"
+)
+
+const (
+	// applicationSnippet is a non-compiled regexp that can be composed with
+	// other snippets to form a valid application regexp.
+	applicationSnippet = "(?:[a-z][a-z0-9]*(?:-[a-z0-9]*[a-z][a-z0-9]*)*)"
+
+	// numberSnippet is a non-compiled regexp that can be composed with other
+	// snippets for validating small number sequences.
+	numberSnippet = "(?:0|[1-9][0-9]*)"
+
+	// unitSnippet is a non-compiled regexp for a valid unit name.
+	unitSnippet = "(" + applicationSnippet + ")/(" + numberSnippet + ")"
+)
+
+const (
+	InvalidUnitName = errors.ConstError("invalid unit name")
+)
+
+var validUnit = regexp.MustCompile("^" + unitSnippet + "$")
+
+// Name represents a units name, used as a human-readable unique identifier.
+type Name string
```

Comment:
OT: I do find it strange that it's name and not ID, but I suspect that's a historic thing.